core/services/log/orm.go

```diff
@@ -54,44 +76,164 @@ func (o *orm) WasBroadcastConsumed(blockHash common.Hash, logIndex uint, jobID i
+	if err != nil {
+		return errors.Wrap(err, "failed to create log broadcast")
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	} else if rowsAffected == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
 func (o *orm) MarkBroadcastConsumed(blockHash common.Hash, blockNumber uint64, logIndex uint, jobID int32, qopts ...postgres.QOpt) error {
 	q := postgres.NewQ(o.db, qopts...)
-	res, err := q.Exec(`INSERT INTO log_broadcasts (block_hash, block_number, log_index, job_id, created_at, consumed, evm_chain_id) VALUES ($1, $2, $3, $4, NOW(), true, $5)`, blockHash, blockNumber, logIndex, jobID, o.evmChainID)
+	res, err := q.Exec(`
+        INSERT INTO log_broadcasts (block_hash, block_number, log_index, job_id, created_at, updated_at, consumed, evm_chain_id)
+		VALUES ($1, $2, $3, $4, NOW(), NOW(), true, $5)
+		ON CONFLICT (job_id, block_hash, log_index, evm_chain_id) DO UPDATE
+		SET consumed = true, updated_at = NOW()
+    `, blockHash, blockNumber, logIndex, jobID, o.evmChainID)
 	if err != nil {
-		return errors.Wrap(err, "while marking log broadcast as consumed")
+		return errors.Wrap(err, "failed to mark log broadcast as consumed")
 	}
 	rowsAffected, err := res.RowsAffected()
 	if err != nil {
-		return errors.Wrap(err, "MarkBroadcastConsumed failed")
+		return err
+	} else if rowsAffected == 0 {
+		return sql.ErrNoRows
 	}
-	if rowsAffected == 0 {
-		return errors.Errorf("cannot mark log broadcast as consumed: does not exist")
+	return nil
+}
+
+func (o *orm) RemoveUnconsumedSetPending(qopts ...postgres.QOpt) (*int64, error) {
+	minUnconsumed, err := o.getUnconsumedMinBlock(qopts...)
+	if err != nil {
+		return nil, err
+	}
+	minPending, err := o.GetBroadcastsPending(qopts...)
+	if err != nil {
+		return nil, err
+	}
+	if minUnconsumed == nil {
+		return minPending, nil
+	}
+	if minPending == nil || *minUnconsumed < *minPending {
+		minPending = minUnconsumed
+		if err := o.SetBroadcastsPending(minPending, qopts...); err != nil {
+			return nil, err
+		}
+	}
+	if err := o.removeUnconsumed(qopts...); err != nil {
+		return nil, err
+	}
+	return minPending, nil
+}
+
+func (o *orm) SetBroadcastsPending(blockNumber *int64, qopts ...postgres.QOpt) error {
+	q := postgres.NewQ(o.db, qopts...)
+	res, err := q.Exec(`
+        INSERT INTO log_broadcasts_pending (evm_chain_id, block_number, created_at, updated_at) VALUES ($1, $2, NOW(), NOW())
+		ON CONFLICT (evm_chain_id) DO UPDATE SET block_number = $3, updated_at = NOW() 
+    `, o.evmChainID, null.IntFromPtr(blockNumber), null.IntFromPtr(blockNumber))
+	if err != nil {
+		return errors.Wrap(err, "failed to set pending broadcast block number")
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	} else if rowsAffected == 0 {
+		return sql.ErrNoRows
+	}
+	return nil
+}
+
+func (o *orm) GetBroadcastsPending(qopts ...postgres.QOpt) (*int64, error) {
+	q := postgres.NewQ(o.db, qopts...)
+	var blockNumber null.Int
+	err := q.QueryRowx(`
```

Comment:
@jkongie would point out this can be done with a `q.Get(&blockNumber, "SELECT...")`